### ISSUE TITLE
lake progress meter displays as garbage symbols on some windows systems

### ISSUE THREAD

**Issue by kim-em** [leanprover/lean4#4291]
I'm not sure the exact parameters of *which* Windows systems are affected by this, but I observed the following output on a user's machine:

```
ΓÜá [2246/2248] Replayed MIL.C11_Integration_and_Measure_Theory.S03_Integration
```

(This is not specific to the progress meter, of course, although it is the first place a new user will encounter it --- Lean error messages that used Unicode symbols from the code are displayed incorrectly as well --- so perhaps the correct resolution is to try to detect this situation, and give the user an actionable information message about what they could do to fix it.)

**Comment by Kha** [comment]
Which terminal is that, VS Code's?

**Comment by semorrison** [comment]
Yes, when I've observed this it has been inside VSCode, and I *think* it was powershell (or at least, the string `PS` appeared in the command prompt).

If anyone can confirm a Windows repro that would be nice.

**Comment by digama0** [comment]
I saw this yesterday on an italian windows laptop on powershell. I'm pretty sure it has to do with the locale setting.

**Comment by tydeu** [comment]
As I use Windows, I noticed this with Lake and investigated this myself a bit earlier. The problem appears to be twofold:

1. [Windows requires](https://stackoverflow.com/a/45622802) that you set the console codepage to UTF8 (via `SetConsoleOutputCP(CP_UTF8);`). Otherwise, it will generally interpret it as ANSI. Windows STL code will also not buffer unicode properly without additional configuration via `setvbuf`. (I am not sure if this applies to use as we are not building with Visual Studio.)
2. We currently [use `fwrite`](https://github.com/leanprover/lean4/blob/390a9a63a24ca28306a60faa0381d4292d23af95/src/runtime/io.cpp#L516) to output strings byte-by-byte. [`fwrite`](https://en.cppreference.com/w/c/io/fwrite) just repeatedly calls `fputc` on each byte. This [will not be printed as Unicode](https://stackoverflow.com/a/45588456) on Windows, as it requires each `fputc` to be a valid codepoint (e.g., via wide characters) for it to render them.

I had attempting to fix this on my TODO list, but have not yet gotten around to it.

**Comment by Kha** [comment]
> [fwrite](https://en.cppreference.com/w/c/io/fwrite) just repeatedly calls fputc on each byte.

Do you have a reference for that? "as if" usually means that it specifically *doesn't* do that.

**Comment by tydeu** [comment]
@Kha I tested it locally with the following program, and you are correct: the `fwrite` does not appear to matter.

**unicode.c**
```c
#include <stdio.h>
#ifdef _WIN32
#include <windows.h>
#endif

int main(int argc, char** argv) {
#ifdef _WIN32
  SetConsoleOutputCP(CP_UTF8);
#endif
  char str[] = "⣾⣷⣯⣟⡿⢿⣻⣽";
  fwrite(str, 1, sizeof(str), stdout);
  fputs("\n", stdout);
  fwrite(str, sizeof(str), 1, stdout);
  fputs("\n", stdout);
  return 0;
}
```

This program produces the following output: (on Windows and Ubuntu):

```
$ clang unicode.c -o unicode && ./unicode.exe 
⣾⣷⣯⣟⡿⢿⣻⣽
⣾⣷⣯⣟⡿⢿⣻⣽ 
```

Without `SetConsoleOutputCP(CP_UTF8);`,  Windows instead produces garbled Unicode:

```
$ clang unicode.c -o unicode && ./unicode.exe 
Γú╛Γú╖Γú»ΓúƒΓí┐Γó┐Γú╗Γú╜
Γú╛Γú╖Γú»ΓúƒΓí┐Γó┐Γú╗Γú╜
```

**Comment by kim-em** [comment]
Still happening for me (Windows under Parallels).

<img width="405" alt="Screenshot 2024-12-06 at 11 02 14 AM" src="https://github.com/user-attachments/assets/836b70fc-98c8-48aa-8284-75ea38ef6b09">
